### PR TITLE
Allow containers to be retried.

### PR DESCRIPTION
**Pull request by shrinandj**
This commit allows `container` templates to be retried. The user
simply adds a 'retries' section to the templated. Currently, this
section only has a 'limit' field that has the number of times the
container should be retried in case of failure.

Subsequent commits will add more features such as retries policies,
retries on specific errors, erc.

When a container template has a retries section, the workflow
controller adds an intermediate node to the workflow. This node
acts as a parent of all the retries.

The argo command line tool's 'get' command correctly shows the
intermediate node and the child nodes along with their pod
names and status'.

Added unit tests that check the various state transitions of the
intermediate node based on the status of the child nodes (which
actually execute the pod).

Testing Done.

* Unit tests succeeded.

* Ran simple workflow that had retries. It completed and status
  of the nodes was correct.

* Ran workflows with parallel steps, each of which had containers
  with retries. The workflow completed correctly and the retries
  were applied to each individual container correctly.

* Ran other workflows which did not have retries. The completed
  correctly.

* `argo get <wf-name>` shows the retries and the pod names.

### CODE REVIEW

File: examples/container_retries.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: container-retries-
+spec:
+  entrypoint: container-retries
+  templates:
+  - name: container-retries
+    retries:
```
Comment (sandeepbhojwani) [inline examples/container_retries.yaml]: should retries be part of a step instead? ex: i have a workflow with multiple steps and some of the steps use the same template but with different retries.

**Comment by edlee2121** [comment]
To clarify, the intermediate node is just an internal control node, not a pod?
The intermediate nodes are not visible to users?

File: examples/container_retries.yaml (hunk 1 above)
Comment (shrinandj) [reply examples/container_retries.yaml]: I agree, there is a use-case for retries with steps too. However, a step can itself be a workflow with multiple steps and that makes the semantics harder to express (e.g. if a step has multiple containers and some of these completed but others didn't, should the step be retried?).

Expressing the retries at the level of leaf node is easier and unambiguous. Hence this implementation has taken that approach as the first step.

As the implementation matures, retries could be added to the steps as well.

**Comment by shrinandj** [comment]
That's right. The intermediate node is an internal node and not a pod. Currently, the output of `argo get` does show this node. E.g. 

```
STEP                              PODNAME                             MESSAGE
 ✖ container-retries-6n4rn
 └---✖ container-retries-6n4rn-0  container-retries-6n4rn-1631060479  failed with exit code 1
 ├---✖ container-retries-6n4rn-1  container-retries-6n4rn-1614282860  failed with exit code 1
 ├---✖ container-retries-6n4rn-2  container-retries-6n4rn-1664615717  failed with exit code 1
 ├---✖ container-retries-6n4rn-3  container-retries-6n4rn-1647838098  failed with exit code 1
 └---✖ container-retries-6n4rn-4  container-retries-6n4rn-1563950003  failed with exit code 1
```

**Comment by edlee2121** [comment]
I guess there is a benefit to showing it as well.

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -247,6 +258,49 @@ func (woc *wfOperationCtx) persistUpdates() {
...
+		}
+
+		if lastChildNode.Successful() {
+			woc.markNodePhase(node.Name, wfv1.NodeSucceeded)
+			return false, nil
+		}
+
+		if !lastChildNode.CanRetry() {
+			woc.log.Infof("Node cannot be retried. Marking it failed")
+			woc.markNodePhase(node.Name, wfv1.NodeFailed)
```
Comment (jessesuen) [inline workflow/controller/operator.go]: Please propagate the failure reason when marking nodes as failed.

File: workflow/controller/operator.go [hunk 3]
```diff
@@ -247,6 +258,49 @@ func (woc *wfOperationCtx) persistUpdates() {
...
+			return false, nil
+		}
+
+		if int32(len(node.Children)) > node.RetryInfo.Limit {
+			woc.log.Infoln("No more retries left. Failing...")
+			woc.markNodePhase(node.Name, wfv1.NodeFailed)
```
Comment (jessesuen) [inline workflow/controller/operator.go]: Please propagate the failure reason when marking nodes as failed.

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -258,18 +312,26 @@ func (woc *wfOperationCtx) podReconciliation() error {
...
 		}
 	}
-	if len(podList.Items) > 0 {
+
+	retriesLeft, err := woc.processNodesWithRetries()
+	if err != nil {
+		return err
+	}
+
+	if retriesLeft || len(podList.Items) > 0 {
```
Comment (jessesuen) [inline workflow/controller/operator.go]: Why do we need to consult retriesLeft to see if we should continue the check for deleted pods? Shouldn't this be left as:
```
if len(podList.Items) > 0 {
```

File: pkg/apis/workflow/v1alpha1/types.go [hunk 5]
```diff
@@ -143,6 +143,8 @@ type Template struct {
 	// before the system actively tries to terminate the pod; value must be positive integer
 	// This field is only applicable to container and script templates.
 	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+
+	RetryInfo *RetryInfo `json:"retries,omitempty"`
```
[the remark this answers is not in the document]
Comment (jessesuen) [reply pkg/apis/workflow/v1alpha1/types.go]: Also, what do you think about RetryStrategy as a struct name vs. RetryInfo. It follows other kubernetes "strategies" structs (e.g. DeploymentStrategy, StatefulSetUpdateStrategy) and describes _*how*_ you want to retry.

[the remark this answers is not in the document]
Comment (shrinandj) [reply pkg/apis/workflow/v1alpha1/types.go]: Changed to `RetryStrategy`.

File: workflow/controller/operator.go [hunk 6]
```diff
@@ -247,6 +258,49 @@ func (woc *wfOperationCtx) persistUpdates() {
 	}
 }
 
+// processNodesWithRetries checks whether any of the failed nodes had retries left and if so,
+// starts a new pod. It returns whether there are retries left or not.
```
[the remark this answers is not in the document]
Comment (shrinandj) [reply workflow/controller/operator.go]: Fixed.

File: pkg/apis/workflow/v1alpha1/types.go [hunk 7]
```diff
@@ -263,6 +265,21 @@ type WorkflowStatus struct {
...
+	for _, node := range wfs.Nodes {
+		if node.RetryInfo != nil && node.RetryInfo.Limit > 0 {
+			nodesWithRetries = append(nodesWithRetries, node)
+		}
+	}
+	return nodesWithRetries
+}
+
+type RetryInfo struct {
+	Limit int32 `json:"limit"`
```
[the remark this answers is not in the document]
Comment (shrinandj) [reply pkg/apis/workflow/v1alpha1/types.go]: Done.

File: workflow/controller/operator.go [hunk 8]
```diff
@@ -247,6 +258,49 @@ func (woc *wfOperationCtx) persistUpdates() {
...
+			return false, fmt.Errorf("Failed to find last child of node " + node.Name)
+		}
+
+		if !lastChildNode.Completed() {
+			// last child node is still running.
+			return false, nil
+		}
+
+		if lastChildNode.Successful() {
+			woc.markNodePhase(node.Name, wfv1.NodeSucceeded)
```
[the remark this answers is not in the document]
Comment (shrinandj) [reply workflow/controller/operator.go]: Good catch. Didn't consider outputs. Copied the outputs to the parent node.

File: workflow/controller/operator.go (hunk 2 above)
Comment (shrinandj) [reply workflow/controller/operator.go]: Done.

File: workflow/controller/operator.go (hunk 3 above)
Comment (shrinandj) [reply workflow/controller/operator.go]: Done.

File: workflow/controller/operator.go (hunk 4 above)
Comment (shrinandj) [reply workflow/controller/operator.go]: Thanks for pointing this out again. At some point, there lower `assessNodeStatus` was marking the intermediate node as failed with a `pod deleted` error and therefore I had to bypass that. No longer needed.

`processNodesWithRetries` only returns an error now.

File: workflow/controller/operator.go [hunk 9]
```diff
@@ -764,6 +879,7 @@ func (woc *wfOperationCtx) markNodePhase(nodeName string, phase wfv1.NodePhase,
 	}
 	woc.wf.Status.Nodes[nodeID] = node
 	woc.updated = true
+	woc.log.Debugf("Marked node %s %s\n", nodeName, phase)
```
[the remark this answers is not in the document]
Comment (shrinandj) [reply workflow/controller/operator.go]: Done.

File: workflow/controller/operator.go [hunk 10]
```diff
@@ -633,12 +699,26 @@ func (woc *wfOperationCtx) deletePVCs() error {
...
+}
+
 func (woc *wfOperationCtx) executeTemplate(templateName string, args wfv1.Arguments, nodeName string) error {
-	woc.log.Debugf("Evaluating node %s: template: %s", nodeName, templateName)
+	woc.log.Infof("Evaluating node %s: template: %s", nodeName, templateName)
 	nodeID := woc.wf.NodeID(nodeName)
 	node, ok := woc.wf.Status.Nodes[nodeID]
 	if ok && node.Completed() {
-		woc.log.Debugf("Node %s already completed", nodeName)
+		woc.log.Infof("Node %s already completed", nodeName)
```
[the remark this answers is not in the document]
Comment (shrinandj) [reply workflow/controller/operator.go]: This is an important message and seems invaluable for debugging given the default logs. But nonetheless, changed it back to debug.

**Comment by jessesuen** [comment]
One more thing -- I think you will need to regenerate the autogenerated pkg/client files since you changed the types. I did not see those files changed.

**Comment by jessesuen** [comment]
Also, can we rename the examples so they are all prefixed by `retries`? This will group the yamls logically by the feature they are demonstrating.

**Comment by shrinandj** [comment]
* Renamed the example files to `retry-container.yaml`, `retry-container-to-completion.yaml` and `retry-with-steps.yaml`.

* Ran `$SRC/hack/update-codegen.sh` to generate new code and committed it. `$SRC/hack/verify-codegen.sh` ran successfully and did not show any errors.

**Comment by gaoning777** [comment]
Does argo schedule an extra pod/container that launches and monitors the container that does the actual work?

**Comment by gaoning777** [comment]
Seems like the retry logic is in the controller: https://github.com/shrinandj/argo/blob/b144283923b23bb8c0707f49295fc7a219acdfc4/workflow/controller/operator.go#L781 instead of in an extra k8s container/pod, right?

**Pull request merged**